backend/src/main/java/kernel360/techpick/feature/folder/service/FolderService.java

```diff
@@ -0,0 +1,122 @@
+import lombok.RequiredArgsConstructor;
+
+@Service
+@RequiredArgsConstructor
+public class FolderService {
+
+	private final FolderProvider folderProvider;
+	private final FolderMapper folderMapper;
+
+	@Transactional
+	public FolderResponse createFolder(Long userId, FolderCreateRequest request) throws ApiFolderException {
+
+		// 생성하려는 폴더의 부모가 본인 폴더인지 검증
+		Folder parent = folderProvider.findById(request.parentId());
+		validateFolderAccess(userId, parent);
+
+		// 생성하려는 폴더 이름이 중복되는지 검증
+		validateDuplicateFolderName(userId, request.name());
+
+		Folder folder = folderProvider.save(folderMapper.createFolder(userId, request, parent));
+
+		return folderMapper.createResponse(folder);
+	}
+
+	@Transactional(readOnly = true)
+	public List<FolderResponse> getFolderListByUserId(Long userId) {
+
+		return folderProvider.findAllByUserId(userId)
+			.stream()
+			.map(folderMapper::createResponse)
+			.toList();
+	}
+
+	@Transactional(readOnly = true)
+	public List<FolderResponse> getFolderListByParentId(Long userId, Long parentId) {
+
+		return folderProvider.findAllByUserIdAndParentId(userId, parentId)
+			.stream()
+			.map(folderMapper::createResponse)
+			.toList();
+	}
+
+	@Transactional
+	public void updateFolder(Long userId, FolderUpdateRequest request) throws ApiFolderException {
+
+		// 변경하려는 폴더가 본인 폴더인지 검증
+		Folder targetFolder = folderProvider.findById(request.id());
+		validateFolderAccess(userId, targetFolder);
+
+		// 변경하려는 폴더의 부모가 본인 폴더인지 검증
+		Folder parent = folderProvider.findById(request.parentId());
+		validateFolderAccess(userId, parent);
+
+		// 수정하려는 폴더 이름이 중복되는지 검증
+		validateDuplicateFolderName(userId, request.name());
+
+		targetFolder.update(request.name(), parent);
+		folderProvider.save(targetFolder);
+	}
+
+	@Transactional
+	public void moveToUnclassified(Long userId, Long folderId) {
+
+		// 이동하려는 폴더가 본인 폴더인지 검증
+		Folder targetFolder = folderProvider.findById(folderId);
+		validateFolderAccess(userId, targetFolder);
+
+		targetFolder.updateParent(folderProvider.findUnclassified(userId));
+		folderProvider.save(targetFolder);
+	}
+
+	@Transactional
+	public void moveToRecycleBin(Long userId, Long folderId) {
+
+		// 이동하려는 폴더가 본인 폴더인지 검증
+		Folder targetFolder = folderProvider.findById(folderId);
+		validateFolderAccess(userId, targetFolder);
+
+		targetFolder.updateParent(folderProvider.findRecycleBin(userId));
+		folderProvider.save(targetFolder);
+	}
+
+	@Transactional
+	public void deleteFolder(Long userId, Long folderId) throws ApiFolderException {
+
+		// 삭제하려는 폴더가 본인 폴더인지 검증
+		Folder targetFolder = folderProvider.findById(folderId);
+		validateFolderAccess(userId, targetFolder);
+
+		// TODO: 삭제된 폴더안의 다른 폴더와 픽들의 부모 설정은 FolderPickFacade에서 미분류로 변경
+		folderProvider.deleteById(folderId);
+	}
+
+	private void validateFolderAccess(Long userId, Folder folder) throws ApiFolderException {
+
+		if (folder == null || Objects.equals(userId, folder.getUser().getId())) {
```

Comment:
p2:
1. folder == null은 access denied 예외와 관련 없음 --> if 문에서 null을 체크할 필요가 없음.
2. 로직상 provider에서 null이 아닌 folder를 제공해줌. 만약 null이 들어왔다면, 그건 개발자가 provider 안쓰고 repository 써서 한 것(개발자 책임)
 
근데 안전을 위해 검증하는 건 필요하니까, if문에서 하지 않고 assertion을 따로 하는게 맞지 않을까.

제가 제안하는 방식은 일단 assert로 적고,

```suggestion
		assert ( folder != null ); // TODO: change to our exception --> ex. Api.assert(folder != null);
```
이렇게 위에 두고, 나중에 리팩토링할 때 우리가 따로 개발자에게 책임을 물을 수 있는 Exception을 추가하면 어떨까

Reply:
수정완료